app/controllers/composer_controller.rb

```diff
@@ -0,0 +1,166 @@
+# frozen_string_literal: true
+
+class ComposerController < ApplicationController
+  requires_login
+
+  def mentions
+    names = params.require(:names)
+    raise Discourse::InvalidParameters.new(:names) if !names.kind_of?(Array) || names.size > 20
+
+    if params[:topic_id].present?
+      topic = Topic.find_by(id: params[:topic_id])
+      guardian.ensure_can_see!(topic)
+    end
+
+    # allowed_names is necessary just for new private messages.
+    allowed_names = if params[:allowed_names].present?
+      params[:allowed_names].split(',') << current_user.username
+    else
+      []
+    end
+
+    user_reasons = {}
+    group_reasons = {}
+    here_count = nil
+
+    users = User
+      .not_staged
+      .where(username_lower: names.map(&:downcase))
+      .index_by(&:username_lower)
+
+    groups = Group
+      .visible_groups(current_user)
+      .where(name: names)
+      .index_by(&:name)
```

Comment:
Since we're on Rails 7 now, you can consider using [load_async](https://api.rubyonrails.org/classes/ActiveRecord/Relation.html#method-i-load_async) but we don't have to do this in this PR. Something to think about when we're executing queries which are independent of each other.

Reply:
I will query the DB only when I need the data anyway so I don't think it makes sense anymore.